Mobile/ShopLens/Droid/Activities/ShoppingCartActivity.cs

```diff
@@ -1,50 +1,171 @@
             addItemButton = FindViewById<Button>(Resource.Id.ShopCartAddItemButton);
             addItemEditText = FindViewById<EditText>(Resource.Id.ShopCartAddItemEditText);
+            recogniseVoice = FindViewById<Button>(Resource.Id.ShopCartRecVoice);
 
-            ArrayAdapter<string> listAdapter = 
+            listAdapter = 
                 new ArrayAdapter<string>(this, Android.Resource.Layout.SimpleListItemChecked, items);
             listView.Adapter = listAdapter;
             listView.ChoiceMode = ChoiceMode.Multiple;
 
-            addItemButton.Click += (sender, e) =>
+            addItemButton.Click += AddTextBoxProductToList;
+            recogniseVoice.Click += RecogniseVoice;
+        }
+
+        private void RecogniseVoice(object sender, EventArgs e)
+        {
+            speechIntent = VoiceRecognizerHelper.SetUpVoiceRecognizerIntent();
+            commandRecognizer.StartListening(speechIntent);
+        }
+
+        public void OnInit([GeneratedEnum] OperationResult status)
+        {
+            // If we get an error, use the default language.
+            if (status == OperationResult.Error)
+            {
+                tts.SetLanguage(Locale.Default);
+            }
+            else if (status == OperationResult.Success)
+            {
+                tts.SetLanguage(Locale.Us);
+            }
+        }
+
+        private void SpeakOut(string message)
+        {
+            if (!string.IsNullOrEmpty(message))
+            {
+                tts.Speak(message, QueueMode.Flush, null, null);
+
+                while (tts.IsSpeaking) { }
+            }       
+        }
+
+        private void AddTextBoxProductToList(object sender, EventArgs e)
+        {
+            AddStringToList(addItemEditText.Text);
+        }
+
+        private void AddStringToList(string text)
+        {
+            if (!string.IsNullOrWhiteSpace(text))
+            {
+                listAdapter.Add(text);
+                prefs.AddString(text);
+            }
+            listAdapter.NotifyDataSetChanged();
+        }
+
+        public void OnResults(Bundle results)
+        {
+            var matches = results.GetStringArrayList(SpeechRecognizer.ResultsRecognition);
+            if (matches.Count > 0)
             {
-                if (!string.IsNullOrWhiteSpace(addItemEditText.Text)){
-                    listAdapter.Add(addItemEditText.Text);
-                    prefs.AddString(addItemEditText.Text);
+                Regex addProductRegex = new Regex(@"^" + @ConfigurationManager.AppSettings["CmdAddCartList"] + @"(?:\s\w+)+");
+
+                if (matches[0] == voiceCartCmd)
+                {
+                    string endMessage = "Voicing of cart complete.";
+
+                    foreach (string item in prefs.GetPreferencesToList())
+                    {
+                        SpeakOut(item);
+                        Thread.Sleep(int.Parse(ConfigurationManager.AppSettings["VoicerPauseTime"]));
```

Comment:
We need to pause between each item so that the user has time to contemplate the results.